targets/mbedos5/tools/generate_pins.py

```diff
@@ -0,0 +1,224 @@
+        print("Fatal: mbed-os directory does not exist.")
+        print("Try running 'make getlibs'")
+        sys.exit(1)
+
+    description = """
+    Generate pins.js for a specified mbed board, using target definitions from the
+    mbed OS source tree.
+    """
+
+    parser = argparse.ArgumentParser(description=description)
+
+    parser.add_argument('board', help='mbed board name')
+    parser.add_argument('-o',
+                        help='Output JavaScript file (default: %(default)s)',
+                        default='js/pins.js',
+                        type=argparse.FileType('w'))
+    parser.add_argument('-c',
+                        help='Output C++ file (default: %(default)s)',
+                        default='source/pins.cpp',
+                        type=argparse.FileType('w'))
+
+    args = parser.parse_args()
+    board_name = args.board.upper()
+
+    target = Target(board_name)
+
+    directory_labels = ['TARGET_' + label for label in target.get_labels()] + target.macros
+
+    targets_dir = os.path.join('.', 'mbed-os', 'hal', 'targets')
+    hal_dir = os.path.join(targets_dir, 'hal')
+
+    pins_file = find_file(hal_dir, directory_labels, 'PinNames.h')
+
+
+    includes = enumerate_includes(targets_dir, directory_labels)
+    defines = list(directory_labels)
+
+    # enumerate pins from PinNames.h
+    pins = enumerate_pins(pins_file, ['./tools'] + list(includes), defines)
+
+    out_file = '\r\n'.join(['var %s = %s;' % pin for pin in pins.iteritems()])
+    args.o.write(out_file)
+
+    LICENSE = '''/* Copyright 2016 ARM, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the \"License\");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an \"AS IS\" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This file is generated by generate_pins.py. Please do not modify.
+ */
+    '''
+
+    COUNT = '''
+unsigned int jsmbed_js_magic_string_count = {};
+    '''.format(len(pins))
+
+    LENGTHS = ',\n    '.join(str(len(name)) for name in pins.iterkeys())
+    LENGTHS_SOURCE = '''
+unsigned int jsmbed_js_magic_string_lengths[] = {
+    %s
+};
+    ''' % LENGTHS
+
+    MAGIC_VALUES = ',\n    '.join(str(value) for value in pins.itervalues())
+    MAGIC_SOURCE = '''
+unsigned int jsmbed_js_magic_string_values[] = {
+    %s
+};
+    ''' % MAGIC_VALUES
+
+    MAGIC_STRINGS = ',\n    '.join('"' + name + '"' for name in pins.iterkeys())
+    MAGIC_STRING_SOURCE = '''
+const char * jsmbed_js_magic_strings[] = {
+    %s
+};
+    ''' % MAGIC_STRINGS
+
+    args.c.write(LICENSE + COUNT + LENGTHS_SOURCE + MAGIC_SOURCE + MAGIC_STRING_SOURCE)
```

Comment:
There is no need to concatenate each string into one before writing it out. We should write out each one with a separate call.